### DISCUSSION TITLE
Provenance Hash Questions

### DISCUSSION THREAD

**Discussion by ghost** [nftchef/art-engine#65] [Help + Q&A]
I have a questions about provenance hash.

Projects such as Bored Ape Yacht Club have a [webpage like this](https://boredapeyachtclub.com/#/provenance). Picture:
<img width="1125" alt="154821685-f279004f-7715-493f-8962-c54fffecd05e-1" src="https://user-images.githubusercontent.com/10701375/155454196-c1efd48b-3340-4207-8304-d47ca09a9088.png">
<img width="1168" alt="Screen Shot 2022-02-23 at 10 58 01 PM" src="https://user-images.githubusercontent.com/10701375/155455358-5b838408-3035-457e-906c-8f71bf1a360c.png">

I'm not sure how one would recreate this Bored Ape Yacht Club webpage when using this library. Some questions I have are:
1. How does one choose a starting index? Is this randomly chosen by us and set here in library as shown [here](https://github.com/nftchef/art-engine/blob/nested-folder-structure/documentation/readme/quickstart.md#start-index) with `const startIndex = 0;`? Some projects like [Invisible Friends](https://invisiblefriends.io/provable-fairness) seem to do complicated things to get a starting index as shown in link.
2. Is this formula `(tokenId + startingIndex) % 10000 → Initial Sequence Index` used at all in this library? Not sure if this is something done somewhere else.
3. I generated the provenance hash with `node utils/provenance.js` using sample artwork. However, I don't see the hashes of individual images. Are these provided?
4. Is the Bored Ape Yacht Club webpage overkill?

**Comment by nftchef** [comment]
Hi, this is a more complex subject.

4. for most projects, yes. Attention to provenance and offset "true randomness" has fallen a bit out of popularity in the last couple of months. It's still a good practice to handle this in the contract from the beginning, but perhaps a page like this is best saved for after minting is closed and if the community asks for it (my opinion)

1. The start index is usually chosen in a single Chainlink VRF call to get a random number (in the smart contract)
2. This formula _can_ be used in the smart contract for convenience, but it generally will happen off-chain on your metadata server (so you won't be using IPFS initially)

3. The generator will output individual hash for each image as long as the option is set to true in config.js. Then, after generation, when running the provenance hash script, it will output provenance.json containing the concatenated string (like BAYC) and the single provenance hash

Hope that helps! I am moving this to a discussion to preserve the conversation for others.

**Reply by ghost** [reply]
@nftchef Thanks for taking time to answer! I see now see the individual hashes for each image along with the concatenated/final provenance hash. 

I'm still a little unsure how to use this library and a smart contract together for provenance hash. Meaning from beginning to launch. 

Could you please outline the steps of how you would launch a collection with provenance hash properly using this library? (this could even be a new tutorial in readme docs). 

**For example, maybe listing out steps such as:**

1. Create start index using Chainlink VRF call. 
2. Change start index in art engine inside of `config.js`
3. Generate images with `npm run generate `
4. Generate provenance hash with  `node utils/provenance.js`
5. Set start index in smart contract to the same one in art engine. This must match!
Your [documentation](https://github.com/nftchef/art-engine/blob/nested-folder-structure/documentation/readme/quickstart.md#start-index) shows this for start index and says to match contract. Still unsure why start index is important for any of this? 
<img width="600" alt="index" src="https://user-images.githubusercontent.com/10701375/155662321-05dd654e-8610-4085-9cbc-5ba6406723df.png">

6. Next step...
7. Next step...

Would you be able to outline correct steps? Above is just example. I could then help to create readme/docs to add to library based on this answer for others. Basically trying to create the best tutorial out there for provenance hash with best practices.

**Reply by nftchef** [reply]
This is a pretty good start! This would definitely be a longer tutorial and one I haven't seen yet

the on-chain part is very simple, you are just adding the provenance hash and offset number to the contract so it is on-chain, trackable if it changes, and publicly verifiable. The rest of the information is displayed offline, like BAYC does, or not depending on a project's interest level.

however you decide to store it, it would be good practice to document _what_ it hashes, e.g., images, or metadata, or both, and note what hashing algorithm you used so others _could_ reconstruct it.
```
    string public PROVENANCE_HASH; // keccak256
```
```solidity
   function setProvenance(string memory _provenance) external onlyOwner {
        PROVENANCE_HASH = _provenance;
    }
   ```
   
Optionally, you can add a require statement to prevent mutating the provenance hash if it is already set.
Another option, if you are really prepared is to set it in the constructor

**Reply by ghost** [reply]
@nftchef In regards to your comment, the main thing I still don't understand is the starting index. Your library has this:
<img width="599" alt="Screen Shot 2022-02-25 at 9 24 42 PM" src="https://user-images.githubusercontent.com/10701375/155825389-2e9b7d1e-4fc3-45f2-9774-2e9143d5ed6f.png">
Notice it says **"Be sure it matches your contract!"**

BAYC has this in their contract which is weird:
```solidity
uint256 public startingIndexBlock;
uint256 public startingIndex;
/**
    * Set the starting index for the collection
    */
function setStartingIndex() public {
    require(startingIndex == 0, "Starting index is already set");
    require(startingIndexBlock != 0, "Starting index block must be set");
    
    startingIndex = uint(blockhash(startingIndexBlock)) % MAX_APES;
    // Just a sanity case in the worst case if this function is called late (EVM only stores last 256 block hashes)
    if (block.number.sub(startingIndexBlock) > 255) {
        startingIndex = uint(blockhash(block.number - 1)) % MAX_APES;
    }
    // Prevent default sequence
    if (startingIndex == 0) {
        startingIndex = startingIndex.add(1);
    }
}
```

Not everyone has this from BAYC though. For example, [Doodles contract](https://etherscan.io/address/0x8a90cab2b38dba80c64b7734e58ee1db38b8992e#code) and [Invisible Friends contract ](https://etherscan.io/address/0x59468516a8259058bad1ca5f8f4bff190d30e066#code) don't have all this complex index code - in fact, they don't have anything in their contract about starting index from what I see. [Doodle's provable fairness webpage](https://docs.doodles.app/provable-fairness) and [Invisible Friends provable fairness webpage](https://invisiblefriends.io/provable-fairness) is also more simple than BAYC. BAYC is overkill it seems.

Sorry if this is lots going back and forth, but since your art engine states **starting index must match contract**, just not sure how to do this correctly (also your example above with Solidity code doesn't have starting index). This is why I tried listing things out step-by-step like a small/quick tutorial above. You mentioned IPFS too, so tricky about steps. If you have time, could you clarify this?

**Reply by ghost** [reply]
@nftchef I know you prob have lots to do, but just wanted to see if you could update on _starting index_ question I posted above?

thank you 🙂

**Reply by nftchef** [reply]
NP! sorry for the delay, sometimes I miss these.

Not all contracts inherit sequential token starting indices, from erc721 enumerable for example where the inherited contract implements the start index and keeps track with ++ on each mint for example.

The note in my generator is aimed (mostly) at people using the original Hashlips sample contract where the token index started at `1`, and for anyone else who _prefers_ the 1 vs. 0 start index. They would end up having issues where the first or last token was missing or greater than/equals operator issues for example.

The difference between the BAYC and doodle/invisible friends implementation is mostly around _how_ and _when_ the shuffling happens. (keeping in mind this is the first time I've looked at the provable fairness for those projects)

From doodles:
> After the finalization of art assets, we shuffle all 10,000 Doodles metadata files
A provenance hash is calculated based on the final shuffled order of metadata files
`32223239cec3be97780fd6a43717feee6dc7d7c2d50fa31ef44d6fcead1a3681`
We send a transaction to the smart contract to commit the provenance hash to the blockchain.
https://etherscan.io/tx/0x7ec645bd154c1cdc2166867c0cdcee829b15d70aa4ed8f60319056de4f48f47b

In the Doodle/IF approach, they're getting the seed number from the community then shuffling the data off-chain, then generating and adding the provenance hash. 

But, (unless I'm missing something) there's still an opportunity where the team knows the order pre-reveal, which is why they need the added step(s) for fairness:
From Doodles
> Mint 504 founder's reserve tokens. Unrevealed metadata is also used here ...[In this order]
> A signed attestation by Poopie, Burnt, and Tulip will be provided promising that the tokens will be distributed in this exact order. No cherry-distributing here!

From IF
>  Additionally no core team members will be allowed to shop tokens on secondary in the pre-reveal period.
> Let list sale be open for 1 hour, then mint 78 tokens in one transaction

They both have to "promise" that there won't be any cheating.

Whereas with the BAYC approach, you first upload the provenance hash, presale (or pre-shuffle), then you are both generating _and_ shuffling (offsetting) on-chain post-sale, so you have to do a little more work to unshuffle and _prove_ the provenance, but then you don't _need_ that layer of trust needed, or "signed promise statements" because it's all provably random on-chain. 

You could also add the same layers of "the team will not buy unrevealed" kind of statements for extra trust/nicety.
